services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,528 @@
+                .orElse(new MongoDbDocumentBuilder())
+                .withTenantId(tenantId)
+                .document();
+
+        final Promise<JsonObject> updateTenantPromise = Promise.promise();
+        final TenantDto newTenantDto = new TenantDto(tenantId, newTenant,
+                new Versioned<>(newTenant).getVersion());
+        mongoClient.findOneAndReplaceWithOptions(config.getCollectionName(), updateTenantQuery,
+                JsonObject.mapFrom(newTenantDto), new FindOptions(), new UpdateOptions().setReturningNewDocument(true),
+                updateTenantPromise);
+
+        return updateTenantPromise.future()
+                .compose(updateResult -> Optional.ofNullable(updateResult)
+                        .map(updated -> {
+                            span.log(String.format("successfully updated tenant [%s]", tenantId));
+                            return Future.<OperationResult<Void>> succeededFuture(OperationResult.ok(
+                                    HttpURLConnection.HTTP_NO_CONTENT,
+                                    null,
+                                    Optional.empty(),
+                                    Optional.of(updateResult.getString(MongoDbDeviceRegistryUtils.FIELD_VERSION))));
+                        })
+                        .orElse(MongoDbDeviceRegistryUtils.checkForVersionMismatchAndFail(
+                                tenantId, resourceVersion, findTenant(tenantId))))
+                .recover((error) -> {
+                    if (error instanceof MongoCommandException) {
+                        logDuplicationError(((MongoCommandException) error).getErrorMessage(), tenantId, span);
+                        return Future.succeededFuture(OperationResult.empty(HttpURLConnection.HTTP_CONFLICT));
+                    } else if (error instanceof ServiceInvocationException) {
+                        return Future.failedFuture(((ServiceInvocationException) error));
+                    }
+                    return Future.failedFuture(new ServerErrorException(HttpURLConnection.HTTP_INTERNAL_ERROR,
+                            String.format("Error modifying tenant [%s].", tenantId)));
+                });
+    }
+
+//
+    /**
+     * {@inheritDoc}
+     * @return A future indicating the outcome of the operation.
+     *         The <em>status</em> will be
+     *         <ul>
+     *         <li><em>204 No Content</em> if the tenant has been removed successfully.</li>
+     *         <li><em>403 forbidden</em> if tenant modification is forbidden due to configuration settings.</li>
+     *         <li><em>404 Not Found</em> if no tenant with the given identifier and version exists.</li>
+     *         <li><em>412 Precondition Failed</em> if the version exists but mismatch the current tenants version.</li>
+     *         </ul>
+     */
+    @Override
+    public Future<Result<Void>> deleteTenant(final String tenantId, final Optional<String> resourceVersion,
```

Comment:
We could now remove the _JavaDoc_ of all overridden methods. With the PR  #1879, _JavaDoc_ of the API interfaces have been updated asking to refer to the API documentation for status codes.

Reply:
done